### ISSUE TITLE
MediaPlayerExtended throws exception with certain videos on certain Samsung devices

### ISSUE THREAD

**Issue by apwillies** [protyposis/ITEC-MediaPlayer#56]
Hi Protyposis! 

First of all - we love your project. We have been looking for a way to seek accurately to a specific frame on Android for about 3 years. We have an app that is designed to provide the Bible in sign language to deaf people around the world, and we need to play from a specific location. So MediaPlayer-Extended is exactly what we need. We do have two issues, however. I'll describe the most serious one here.

**An exception is thrown when playing certain videos on most Samsung devices.** This exception will cause the app to crash. Samsung devices make up 2/3 of the Android ecosystem. So this issue is a showstopper for us. Are you able to take a look at this issue to see what can be done? If you need me to provide more information to help with this, please let me know.

Thanks a million in advance!
Paul

-----

Prequisites:
**1. Device:** You will need to run this on a Samsung device. We have reproduced on the following Samsung devices:
Samsung Galaxy Tab 3 10", Android 4.2.2
Samsung Galaxy S3, Android 4.3
Samsung Galaxy Note 8", Android 4.4.2
Samsung Galaxy Note 3, Android 4.4.2
Samsung Galaxy S5, Android 5.1.1

We were to successfully play these videos on the following Samsung devices:
Samsung Galaxy Tab 4, Android 4.1.2
Samsung Galaxy Note 10", Android 4.1.2

All other devices play these videos just fine. We have Motorola, Lenovo, NOOK, and various Nexus devices.

**2. Video:** A specific set of videos fail. These videos have been produced using FCP-X with out-the-box settings. There are literally thousands of these videos that are out there 'in the wild.' These videos have an additional 'video' track with thumbnails to identify each chapter marker. It is possible this is causing the issue. Note that the problem videos play fine on the stock player and all other video-playing apps on all Samsung devices. They also play fine with MediaPlayerExtended on non-Samsung devices. Our older videos were produced with FCP-7 and they did not have any thumbnails for the chapter markers. These older videos play fine using MediaPlayerExtended on all devices.

We created a test video in FCP-X with similar settings and this one also throws the same exception. Here's the link to the [test video](https://1drv.ms/v/s!ApEGFnZ-yg0eg6s148E8HWFD6pyapA). Again, the problem only exhibits on the Samsung devices - other devices work fine. We made a second version with no chapter markers and therefore no additional 'video' thumbnail track and uploaded it [here](https://1drv.ms/v/s!ApEGFnZ-yg0eg6s2vgsrkvB6QlRvvA). This one plays fine. 

**To reproduce,** use the MediaPlayerDemo project: 

1. Download this [video file](https://download-a.akamaihd.net/files/media_bible/1e/nwt_24_Jer_ASL_01_r480P.m4v) or [the test file](https://1drv.ms/v/s!ApEGFnZ-yg0eg6s148E8HWFD6pyapA)
2. Run the MediaPlayerDemo project and tap the "Pick a file" button
3. Pick the file just downloaded
4. Tap the "VideoView" button <-- **The app will crash**

Here's the output from the "Console" window in Android Studio:
```
D/MainActivity: opening video picker...
D/MainActivity: onActivityResult REQUEST_LOAD_VIDEO
D/ProgressBar: setProgress = 0
D/ProgressBar: setProgress = 0, fromUser = false
D/ProgressBar: mProgress = 0mIndeterminate = false, mMin = 0, mMax = 100
D/ProgressBar: setProgressDrawable mProgressDrawable = null, d = android.graphics.drawable.LayerDrawable@425f8c38needUpdate = false
D/ProgressBar: setProgressDrawable drawableHeight = 32
D/ProgressBar: setProgress = 0
D/ProgressBar: setProgress = 0, fromUser = false
D/ProgressBar: mProgress = 0mIndeterminate = false, mMin = 0, mMax = 100
D/AbsSeekBar: AbsSeekBar Constructor: misSeebarAnimationAvailable = false
D/ProgressBar: updateDrawableBounds: left = 0
D/ProgressBar: updateDrawableBounds: top = 0
D/ProgressBar: updateDrawableBounds: right = 152
D/ProgressBar: updateDrawableBounds: bottom = 152
D/MediaPlayer: setVideoRenderTimingMode AUTO
E/SampleTable: mNumSyncSamples error
D/MediaPlayer: {max-input-size=1572864, height=480, csd-0=java.nio.ByteArrayBuffer[position=0,limit=39,capacity=39], mpx-dar=1.7791667, width=854, durationUs=323323000, csd-1=java.nio.ByteArrayBuffer[position=0,limit=8,capacity=8], mime=video/avc, isDMCMMExtractor=1}
D/AsyncDataSource: AsyncDataSource(0x5fe5bf08)::ConstructL bufferSize 1048576 numBuffers 3 readSize 1048576 readAhead 2097152 iAsyncReadThreadState=1
I/AsyncDataSource: AsyncDataSource(0x5fe5bf08)::setRange offset(0x2bd) size(0xca06) iMaxPosition(0xccc3)
D/AsyncDataSource: AsyncDataSource(0x5fe5c940)::ConstructL bufferSize 524288 numBuffers 6 readSize 524288 readAhead 2621440 iAsyncReadThreadState=1
I/AsyncDataSource: AsyncDataSource(0x5fe5c940)::setRange offset(0x43324) size(0x0) iMaxPosition(0x0)
D/MediaPlayer: {max-input-size=1572864, durationUs=323323000, isDMCMMExtractor=1, mime=text/3gpp-tt}
E/SampleTable: mNumSyncSamples error
A/Utils: frameworks/av/media/libstagefright/Utils.cpp:105 CHECK(meta->findInt32(kKeyWidth, &width)) failed.
A/libc: Fatal signal 11 (SIGSEGV) at 0xdeadbaad (code=1), thread 10915 (Thread-6475)

```

**Comment by velocitysystems** [comment]
@protyposis I can also confirm this issue and am particularly interested as it affects playback on a large number of Samsung devices. Please see below for further investigation.

- Test File: "nwt_24_Jer_ASL_01_r480P.m4v"
- Device 1: Samsung Galaxy Tab (SM-T310), Android 4.4.2. **Same crash as reported.**
- Device 2: Google Nexus 7 (2012 model), Android 5.1.1. **No crash.**
- Device 3: Google Nexus 7 (2013 model) , Android 6.0.1. **No crash.**

Testing with the first video file supplied by @apwillies, Device 1 output shows a SIGSEGV is being thrown due to a fatal assertion in `frameworks/av/media/libstagefright/Utils.cpp:105` when trying to retrieve the MediaFormat object for the third track (index=2). Within MediaPlayerExtended, the call site is the `getTrackFormat(int index)` method in `MediaExtractor.java:180`.

Inspecting the file with _Media-Info_ on OSX shows:
- Stream 0 (Track 1): Type=video, Codec=h264, FourCC=avc1
- Stream 1 (Track 2): Type=subtitle
- Stream 2 (Track 3): Type=video, Codec=mjpeg, FourceCC=jpeg

Inspecting the file with _MP4 Explorer_ on Windows shows:
- Track 1: Flags=1 ENABLED, Type=Video, Sample Count=9690
Sample Description={Coding=avc1, Width=854, Height=480}
- Track 2: Flags=0, Type=Text, Sample Count=21, No Sample Description
- Track 3: Flags=0, Type=Video, Sample Count=21, No Sample Description

Debugging the demo project on Device 1 using Android Studio, the fatal crash occurs on `MediaExtractor.java:180` when index=2. When debugging on Device 2 or Device 3, no crash occurs.

When debugging on Device 2 and 3, I noticed that the hash map of keys returned in the MediaFormat object for the third track does not contain a width or height key. This would seem to be consistent with the crash behaviour due to an assertion failure at `CHECK(meta->findInt32(kKeyWidth, &width))' in the core libstagefright utility method.

It would appear then that affected Samsung devices are using a modified version of `libstagefright` and potentially handling the CHECK assertions in a different manner.

It is worth noting that a number of Stagefright exploits were discovered in 2015 and in response  Samsung/Google both patched libstagefright to fix the reported CVEs. [See - Arstechnica Blog 2015/08](https://arstechnica.co.uk/security/2015/08/google-pushes-update-for-critical-android-bug-but-wont-say-if-its-fixed/) Interestingly, another user reported a similar issue on StackOverflow a year ago and wondered also if it may be related to the Stagefright exploit [See - MediaExtractor causes SIGABRT on getTrackFormat()](http://stackoverflow.com/questions/34067049/mediaextractor-causes-sigabrt-on-gettrackformat)

Note: These fatal `CHECK` assertions were subsequently removed in the Android master branch on Tue Jan 26 2016. [See DIFF of Utils.cpp](https://android.googlesource.com/platform/frameworks/av/+/f4101f746a4a2c407a220dd61860a2b742eade51%5E%21/#F1) They have been removed in favour of a more graceful exception handling path.

**Overview**
Given the problem is that `libstagefright` is causing the `SIGABRT` due to the assertion failure triggered by the `CHECK` macro inside `Utils.cpp`, the only solution within MediaPlayerExtended would be to **prevent the MediaExtractor `getTrackFormat(index)` method from being called for the third track.**

Looking at the available API methods with Android's MediaExtractor class, I don't believe there is any simple, alternative method to get the track codec information and skip calling `getTrackFormat(index)` if the codec is MJPEG or if the width/height are missing. In fact, the `getTrackFormat` method exists for this very purpose.

**Possible Workaround**
It may be possible to modify `MediaPlayer.java:220-236` to **break out of the for loop once a match is found for `mVideoTrackIndex`**, this will prevent getTrackFormat(i) from being called for subsequent tracks and resolve this issue in a majority of cases. However, there are some negative side effects to this workaround (see comments below).

**Comment by velocitysystems** [comment]
Proof-of-concept fix below demonstrates the proposed workaround.
However there are a couple of limitations:
1. **Track order.** Workaround relies on the assumption that the valid video track will be before the MJPEG track. May still throw a fatal exception under some conditions. E.g.
If Track 1 = H264, Track 2 = Subtitle (or audio), Track 3 = MJPEG, this is OK.
If Track 1 = Subtitle (or audio), Track 2 = H264, Track 3 = MJPEG, this is OK.
If Track 1 = H264, Track 2 = MJPEG, Track 3 = subtitle (or audio), this will throw.
If Track 1 = MJPEG, Track 2= H264, this will throw etc.

2. **Internal audio track will be ignored if after the video track.** Workaround relies on the assumption that the audio track is located _before_ the valid video track. Note: Will not affect case where a separate audio file is available with its own audio extractor. E.g.
If Track 1 = Audio, Track 2 = H264, this is OK.
If Track 1 = H264, Track 2 = Audio, audio will be ignored.

- Source file is `MediaPlayer.java`, method is `setDataSource(MediaSource source)`
- Track detection loop modified to break once a match is found for video track index. This prevents subsequent (unnecessary) calls to `getTrackFormat` leading to the fatal exception on Samsung devices.

- `break` added at Line 229

        for (int i = 0; i < mVideoExtractor.getTrackCount(); ++i) {
            MediaFormat format = mVideoExtractor.getTrackFormat(i);
            Log.d(TAG, format.toString());
            String mime = format.getString(MediaFormat.KEY_MIME);
            if (mVideoTrackIndex < 0 && mime.startsWith("video/")) {
                mVideoExtractor.selectTrack(i);
                mVideoTrackIndex = i;
                mVideoFormat = format;
                mVideoMinPTS = mVideoExtractor.getSampleTime();
                break;
            } else if (mAudioExtractor == null && mAudioTrackIndex < 0 && mime.startsWith("audio/")) {
                mVideoExtractor.selectTrack(i);
                mAudioTrackIndex = i;
                mAudioFormat = format;
                mAudioMinPTS = mVideoExtractor.getSampleTime();
                mAudioExtractor = mVideoExtractor;
            }
        }

Tested and verified the fix on Device 1: Samsung Galaxy Tab (SM-T310), Android 4.4.2.
**No crash** when playing test file supplied by @apwillies.

**Comment by apwillies** [comment]
@mattrichnz Do you mind putting up a PR for @protyposis to consider?

**Comment by velocitysystems** [comment]
@protyposis @apwillies Proof of concept fix is provided as a sample workaround, however I would recommend spending more time investigating a robust solution that will work for all media, irregardless of track ordering.

A couple of options that may be worth considering:

**Option 1**
**Use `MediaPlayer.TrackInfo` to identify the valid video track.** [See API Doc](https://developer.android.com/reference/android/media/MediaPlayer.TrackInfo.html)
Given the fatal crash does not occur when using the standard Android `MediaPlayer` class, it may be possible to use this to identify the valid video track rather than iterating through the tracks.

Note: I have tested this with the file supplied by @apwillies and it correctly identifies the H.264 track as `MEDIA_TRACK_TYPE_VIDEO` and the MJPEG track as `MEDIA_TRACK_TYPE_UNKNOWN`. With this information, the track discovery loop at `MediaPlayer.java:220-236` could be removed.

MediaPlayerExtended would first need to create an instance of MediaPlayer to retrieve the track information during initialization, however this could then be destroyed. If this is not desirable for all use-cases, an overloaded constructor could make this available as a custom option.

**Option 2**
**Bundle a custom version of `libstagefright` with MediaPlayerExtended.**
Specific calls could then be made using JNI invoke using this known-good version of libstagefright to read the track information.

**Comment by protyposis** [comment]
Thank you very much for the detailed report! 

I'm still not sure about how to best fix this, but I'd like to avoid @mattrichnz's fix due to the pointed out limitations. I think I will go the suggested `MediaPlayer.TrackInfo` route and limit it to Samsung devices and a range of Android versions that still needs to be exactly determined (4.2 up to 5.x or 6.x). 

This will have two disadvantages:
 - remote files will be requested twice, first by the `MediaPlayer` to determine the tracks, then by `MediaPlayer-Extended` to load the media
 - `setDataSource` will take much longer due to the sync `MediaPlayer.prepare()` call... this can lead to lags in the UI when the MediaPlayer is used on the main thread

Especially the second disadvantage is unfortunate, but all other decent fixes would require too much of an effort.

**Comment by apwillies** [comment]
Thanks for the investigation @mattrichnz and the response @protyposis.

**Comment by velocitysystems** [comment]
Thanks @protyposis. I like your proposed suggestion of programatically detecting the device manufacturer/API level and then only using `MediaPlayer.TrackInfo` for these devices. I think that will work well, at least slightly increased latency on Samsung devices is better than a crash!

Look forward to seeing your progress.

**Comment by protyposis** [comment]
The `MediaPlayer.TrackInfo` workaround is working, but unfortunately only for local files. It seems like Samsung's MediaPlayer implementation cannot fetch streams via `http` and always throws the following exception, on which I cannot find anything helpful on the internet:

```
path is null
setDataSource IOException happend : 
java.io.FileNotFoundException: No content provider: http://some.url/to/a/file.mp4
    at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:761)
    at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:665)
    at android.media.MediaPlayer.setDataSource(MediaPlayer.java:960)
```

So it unfortunately seems that this approach won't be an option for a workaround, as I think online stream support is important.

**Comment by velocitysystems** [comment]
Thanks Mario.

Would you be happy to share your work in progress on a remote branch so I can take a look?

In the interim, would it be possible to introduce this as a configuration option or optional ctor argument in MediaPlayerExtended? At least if only for local files until another solution can be found?

I am also interested in a solution which would work for both local and streaming files on Samsung devices.

**Comment by protyposis** [comment]
I found a nice solution, release will follow soon. I figured that this bug is too specific and the workaround too inadequate to solve it directly in MP-X's core, so the solution is the addition of new methods that take track parameters as arguments:

 * `mediaPlayer.setDataSource(MediaSource source, int videoTrackIndex, int audioTrackIndex)`
 * `videoView.setVideoSource(MediaSource source, int videoTrackIndex, int audioTrackIndex)`

The parameters can either be an actual track index, `MediaPlayer.TRACK_INDEX_AUTO` to automatically select the first fitting track index, or `MediaPlayer.TRACK_INDEX_NONE` to not select a track. With this approach, i hand the responsibility to avoid this Samsung bug by determining the track indices through other means (e.g. Android's MediaPlayer) over to the library user :) Sample code for the workaround will follow with the release.

In your case, you are in control over the media files and know which tracks they contain, so you can easily bypass the bug by setting the data source with `setDataSource(source, 0, MediaPlayer.TRACK_INDEX_NONE)` or more generically `setDataSource(source, MediaPlayer.TRACK_INDEX_AUTO, MediaPlayer.TRACK_INDEX_NONE)`.

**Comment by apwillies** [comment]
Thanks for tackling this one! What you say sounds logical. We have many generations of videos, and it's hard to say with certainty that they will always follow a certain pattern. But I'm hopeful this solution will be better than the workaround we have in place right now. I expect @mattrichnz will take a look soon.

**Comment by protyposis** [comment]
Sample code for using Android's MediaPlayer to determine track indices and work around the Samsung crash is available here: https://gist.github.com/protyposis/a9dc703a9107222a09f40dd0b2afff6d

**Comment by protyposis** [comment]
Published as `v4.3.0`. Hope that helps.

**Comment by velocitysystems** [comment]
Wonderful, many thanks Mario.
Look forward to working with the 4.3.0 release.
